### PR TITLE
a stronger version of Where2comm

### PR DESCRIPTION
**Pull request by Little-Podi**
This version should have stronger performance. I don't remember the best configutation, but here are some configurations can be tried to see how it goes:
- Set **where2comm_fusion.fully** to **True** during training and **False** during inference (pruning the communication graph by a pre-defined threshold)
- Set **where2comm_fusion.fully** to **False** during training (random sampling communication ratios, I haven't implemented the incremental learning as the original paper)
- Set **where2comm_fusion.multi_scale** to **False** (previous version has bugs in single-scale fusion, and I have corrected it)

### CODE REVIEW

**Comment by DerrickXuNu** [comment]
Quick question: When we set fully to true, what is the compression rate? Is it something random?

File: opencood/hypes_yaml/point_pillar_fcooper.yaml [hunk 1]
```diff
@@ -81,7 +81,7 @@ model:
       upsample_strides: [1, 2, 4]
       num_upsample_filter: [128, 128, 128]
     shrink_header:
-      kernal_size: [ 1 ]
```
Comment (DerrickXuNu) [inline opencood/hypes_yaml/point_pillar_fcooper.yaml:84]: I know there is a typo here. But all the pretrained models yaml files are using this, it would be a little complicated to modify all of those pretrain models' yaml file. Would you mind change back to the 'kernal_size'? Thanks!

File: opencood/hypes_yaml/point_pillar_where2comm.yaml [hunk 2]
```diff
@@ -1,73 +1,72 @@
 name: point_pillar_where2comm
 root_dir: '/home/cav/data/v2xset/train'
-validate_dir: '/home/cav/data/v2xset//test'
+validate_dir: '/home/cav/data/v2xset/test'
 
 
 wild_setting:
-  async: false
+  seed: 42
```
Comment (DerrickXuNu) [inline opencood/hypes_yaml/point_pillar_where2comm.yaml:7]: I used seed 20 in V2X-ViT paper, so let's keep the same to be consistent with the paper for reproducible results

File: opencood/hypes_yaml/point_pillar_where2comm.yaml [hunk 3]
```diff
@@ -1,73 +1,72 @@
...
+  data_size: 1.06  # Mb
+  transmission_speed: 27  # Mbps
+  backbone_delay: 10  # ms
 
-yaml_parser: "load_point_pillar_params"
+yaml_parser: 'load_point_pillar_params'
 train_params:
-  batch_size: &batch_size 4
-  epoches: &epoches 90
+  batch_size: &batch_size 5
```
Comment (DerrickXuNu) [inline opencood/hypes_yaml/point_pillar_where2comm.yaml:20]: would batch size be an even number better for GPU optimization?

File: opencood/models/sub_modules/downsample_conv.py [hunk 4]
```diff
@@ -33,7 +33,7 @@ def __init__(self, config):
         self.layers = nn.ModuleList([])
         input_dim = config['input_dim']
 
-        for (ksize, dim, stride, padding) in zip(config['kernal_size'],
+        for (ksize, dim, stride, padding) in zip(config['kernel_size'],
```
Comment (DerrickXuNu) [inline opencood/models/sub_modules/downsample_conv.py:36]: let's change this back to the wrong typo version. It is just because all the previous pretrained model use this, the users would have trouble to run the old models

File: opencood/models/point_pillar_where2comm.py [hunk 5]
```diff
@@ -95,17 +95,16 @@ def forward(self, data_dict):
...
             spatial_features_2d = self.naive_compressor(spatial_features_2d)
 
         if self.multi_scale:
+            # Bypass communication cost, communicate at high resolution, neither shrink nor compress
             fused_feature, communication_rates = self.fusion_net(batch_dict['spatial_features'],
                                                                  psm_single,
                                                                  record_len,
                                                                  pairwise_t_matrix,
                                                                  self.backbone)
-            # Down-sample feature to reduce memory
```
Comment (DerrickXuNu) [inline opencood/models/point_pillar_where2comm.py:106]: We need to keep this. Because all the models in V2X-ViT paper shrinked the feature, which can highly influence the performance. We need to make the comparison fair and thus keep this option

**Comment by Little-Podi** [comment]
Where2comm reduces the communication cost by selecting spatially sparse yet perceptually critical areas. **where2comm_fusion.fully** denotes whether we keep the fully-connected communication graph or prune it. In the updated implementation, it works like this:
- During training, if **where2comm_fusion.fully** is **True**, then the model will be always trained with 100% communication without any communication saving. Otherwise, the pruning ratio will be randomly selected from 0% to 100%.
- During inference, if **where2comm_fusion.fully** is **True**, then the agent will send the whole feature map to the ego agent. Otherwise, it will send partial features according to the confidence map and the pre-defined threshold.

**Review by DerrickXuNu — COMMENTED** [review]
Generally lgtm, but there are some points need to be addressed before merging.

**Comment by DerrickXuNu** [comment]
> Where2comm reduces the communication cost by selecting spatially sparse yet perceptually critical areas. **where2comm_fusion.fully** denotes whether we keep the fully-connected communication graph or prune it. In the updated implementation, it works like this:
> 
> * During training, if **where2comm_fusion.fully** is **True**, then the model will be always trained with 100% communication without any communication saving. Otherwise, the pruning ratio will be randomly selected from 0% to 100%.
> * During inference, if **where2comm_fusion.fully** is **True**, then the agent will send the whole feature map to the ego agent. Otherwise, it will send partial features according to the confidence map and the pre-defined threshold.

Thanks for the clarification. Do you know what kind of threshold should we set to be the same as the v2x-vit paper, e.g., 32x compression rate?

File: opencood/hypes_yaml/point_pillar_fcooper.yaml (hunk 1 above)
Comment (Little-Podi) [reply opencood/hypes_yaml/point_pillar_fcooper.yaml:84]: Sure, sure. No problem.

File: opencood/hypes_yaml/point_pillar_where2comm.yaml (hunk 2 above)
Comment (Little-Podi) [reply opencood/hypes_yaml/point_pillar_where2comm.yaml:7]: You are right. Keep 20 for fair comparison.

File: opencood/hypes_yaml/point_pillar_where2comm.yaml (hunk 3 above)
Comment (Little-Podi) [reply opencood/hypes_yaml/point_pillar_where2comm.yaml:20]: I think both even and odd are OK. You can set one that fits the GPU memory. According to my reproduction, a relatively large batch size seems to be better in performance.

File: opencood/models/sub_modules/downsample_conv.py (hunk 4 above)
Comment (Little-Podi) [reply opencood/models/sub_modules/downsample_conv.py:36]: Haha. You may find that I couldn't help myself to change these minor things. Sorry about that. For a widely used project, we shouldn't change these.

File: opencood/models/point_pillar_where2comm.py (hunk 5 above)
Comment (Little-Podi) [reply opencood/models/point_pillar_where2comm.py:106]: Oh, I forgot to mention that. Honestly, I think Where2comm has something confusing in its multi-scale implementation. You will find that it re-extracts the features in multi-scale implementation. It makes the order of **shrink**, **compress**, **fuse** somewhat inappropriate. In my understanding, **shrink** and **compress** should be accomplished before communication. If we add the **shrink** here, it will be done after the feature fusion at the ego agent side. I'm not sure if it's appropriate. Please help me to determine this part. Also, I suggest to use single-scale version.

**Comment by Little-Podi** [comment]
I think 32x will be fine. In Where2comm, they didn't use channel-wise compression to save the communication cost, so I didn't try it.

File: opencood/hypes_yaml/point_pillar_where2comm.yaml (hunk 3 above)
Comment (DerrickXuNu) [reply opencood/hypes_yaml/point_pillar_where2comm.yaml:20]: ok, sounds good to me

File: opencood/models/sub_modules/downsample_conv.py (hunk 4 above)
Comment (DerrickXuNu) [reply opencood/models/sub_modules/downsample_conv.py:36]: No worry, I can totally understand. If you were in the project before, these errors won't happen at all

File: opencood/models/point_pillar_where2comm.py (hunk 5 above)
Comment (DerrickXuNu) [reply opencood/models/point_pillar_where2comm.py:106]: Yeah, I can understand your concern.  What if we shrink first and then fuse them?

**Comment by DerrickXuNu** [comment]
> I think 32x will be fine. In Where2comm, they didn't use channel-wise compression to save the communication cost, so I didn't try it.

I may not ask the question clearly. Other models by default use 32x compression, which will impact the performance. To make fair comparison, where2comm should do the same thing. But it seem they already have compression inside the algorithm, so I am not sure how much compression they have by default/

File: opencood/models/point_pillar_where2comm.py (hunk 5 above)
Comment (Little-Podi) [reply opencood/models/point_pillar_where2comm.py:106]: Sure. That's reasonable.

**Comment by Little-Podi** [comment]
For a pair comparison, you can use 32x. It will make the channel dimension of the transmitted feature map the same as other models. Besides, in my practice, the threshold 0.01 already filter out a large portion of the feature map, which further reduce the communication cost.

**Comment by Little-Podi** [comment]
Oh, I got your point. Because the "compression" mechanism in Where2comm is based on a threshold, it cannot precisely control the ratio such as 1/32 or something. By default, they use 0.01.

**Comment by Little-Podi** [comment]
You can tune this threshold after the model is trained to make the compression rate to be roughly 32x. But the exact rate will be slightly different between different inputs and datasets.

**Comment by DerrickXuNu** [comment]
Thanks for your reply. I think everything looks good to me now except the "kernel" typo. Could you revert back and update it so I can merge yours into my branch?

**Review by DerrickXuNu — APPROVED** [review]
I will merge the pr and modify the typo on my side

**Comment by Little-Podi** [comment]
Sorry, I was preparing a tutorial yesterday. Thanks for your help, anyway.
